internal/service/tesla.go

```diff
@@ -0,0 +1,129 @@
+package service
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"github.com/DIMO-Network/shared/pkg/cipher"
+	"github.com/DIMO-Network/shared/pkg/db"
+	"github.com/DIMO-Network/tesla-oracle/internal/config"
+	dbmodels "github.com/DIMO-Network/tesla-oracle/models"
+	"github.com/rs/zerolog"
+	"regexp"
+	"strconv"
+)
+
+type TeslaService struct {
+	settings *config.Settings
+	logger   *zerolog.Logger
+	Cipher   cipher.Cipher
+	pdb      *db.Store
+}
+
+func NewTeslaService(settings *config.Settings, logger *zerolog.Logger, cipher cipher.Cipher, pdb *db.Store) *TeslaService {
+	return &TeslaService{
+		settings: settings,
+		logger:   logger,
+		Cipher:   cipher,
+		pdb:      pdb,
+	}
+}
+
+type VirtualKeyStatus int
+
+const (
+	Incapable VirtualKeyStatus = iota
+	Paired
+	Unpaired
+)
+
+// GetVehicleByVIN retrieves a vehicle by its VIN.
+func (ts *TeslaService) GetVehicleByVIN(ctx context.Context, logger *zerolog.Logger, pdb *db.Store, vin string) (*dbmodels.SyntheticDevice, error) {
+	tx, err := pdb.DBS().Writer.BeginTx(ctx, &sql.TxOptions{Isolation: sql.LevelReadCommitted})
+	if err != nil {
+		logger.Error().Err(err).Msgf("Failed to begin transaction for vehicle %s", vin)
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			if rbErr := tx.Rollback(); rbErr != nil {
+				logger.Error().Err(rbErr).Msgf("GetVehicleByVIN: Failed to rollback transaction for vehicle %s", vin)
+			}
+		} else {
+			if cmErr := tx.Commit(); cmErr != nil {
+				logger.Error().Err(cmErr).Msgf("GetVehicleByVIN: Failed to commit transaction for vehicle %s", vin)
+			}
+		}
+	}()
+
+	sd, err := dbmodels.SyntheticDevices(
+		dbmodels.SyntheticDeviceWhere.Vin.EQ(vin)).One(ctx, tx)
```

Comment:
What happens when there are two synthetic devices with the same VIN? right now there is nothing that says VIN => SD is 1:1